tiledb/api/src/array/enumeration/mod.rs

```diff
@@ -0,0 +1,529 @@
+                self.capi(),
+                data.as_ptr() as *const std::ffi::c_void,
+                std::mem::size_of_val(data) as u64,
+                offsets_ptr as *const std::ffi::c_void,
+                offsets_len,
+                &mut c_new_enmr,
+            )
+        };
+
+        if res == ffi::TILEDB_OK {
+            Ok(Enumeration {
+                context: self.context,
+                raw: RawEnumeration::Owned(c_new_enmr),
+            })
+        } else {
+            Err(self.context.expect_last_error())
+        }
+    }
+}
+
+impl<'ctx> Debug for Enumeration<'ctx> {
+    fn fmt(&self, f: &mut Formatter) -> FmtResult {
+        let name = self.name().map_err(|_| fmt::Error)?;
+        let dtype = self.datatype().map_err(|_| fmt::Error)?;
+
+        let dtype_string =
+            dtype.to_string().unwrap_or("<unknown datatype>".to_owned());
+        let cell_val_num = self.cell_val_num().map_err(|_| fmt::Error)?;
+        let ordered = self.ordered().map_err(|_| fmt::Error)?;
+
+        let json = json!({
+            "name": name,
+            "datatype": dtype_string,
+            "cell_val_num": cell_val_num,
+            "ordered": ordered,
+            "values": [], // TODO: Render values
+        });
+        write!(f, "{}", json)
+    }
+}
+
+impl<'c1, 'c2> PartialEq<Enumeration<'c2>> for Enumeration<'c1> {
+    fn eq(&self, other: &Enumeration<'c2>) -> bool {
+        let names_match = match (self.name(), other.name()) {
+            (Ok(mine), Ok(theirs)) => mine == theirs,
+            _ => false,
+        };
+        if !names_match {
+            return false;
+        }
+
+        let types_match = match (self.datatype(), other.datatype()) {
+            (Ok(mine), Ok(theirs)) => mine == theirs,
+            _ => false,
+        };
+        if !types_match {
+            return false;
+        }
+
+        let cell_val_num_match =
+            match (self.cell_val_num(), other.cell_val_num()) {
+                (Ok(mine), Ok(theirs)) => mine == theirs,
+                _ => false,
+            };
+        if !cell_val_num_match {
+            return false;
+        }
+
+        let ordered_match = match (self.ordered(), other.ordered()) {
+            (Ok(mine), Ok(theirs)) => mine == theirs,
+            _ => false,
+        };
+        if !ordered_match {
+            return false;
+        }
+
+        let data_match = match (self.data(), other.data()) {
+            (Ok(mine), Ok(theirs)) => mine == theirs,
+            _ => false,
+        };
+        if !data_match {
+            return false;
+        }
+
+        let offsets_match = match (self.offsets(), other.offsets()) {
+            (Ok(Some(mine)), Ok(Some(theirs))) => mine == theirs,
+            (Ok(None), Ok(None)) => true,
+            _ => false,
+        };
+        if !offsets_match {
+            return false;
+        }
+
+        true
+    }
+}
+
+pub struct Builder<'ctx> {
+    context: &'ctx Context,
+    name: String,
```

Comment:
This is a different approach than most of the other Builders which allocate the thing right away and then call the underlying tiledb functions.  I see that it is necessary here because of the way the allocation function looks.  Maybe it would be nice to do this for the other Builders too though (not in this PR, if ever, obviously) - so all the setter methods succeed and then `build` does all the C API work.

Reply:
Huh. Fair point. Constructor args is actually easy enough since I should be able to do that with a single lifetime parameter. I'll give that a quick whirl and if it works out then great. But if the compiler yells too loudly I'll leave it as is.